File: src/automata/DFA.java
package automata;

import java.util.List;
import java.util.Map;

public class DFA {

    private final String name;
    private final List<String> alphabet;
    private final Map<State, Map<String, State>> graph;
    private final State initialState;
    private final List<State> finalStates;

    public DFA(String name, List<String> alphabet, Map<State, Map<String, State>> graph, State initialState, List<State> finalStates) {
        this.name = name;
        this.alphabet = alphabet;
        this.graph = graph;
        this.finalStates = finalStates;
        this.initialState = initialState;
    }

    public boolean doesAcceptString(String w) {

        State currentState = this.initialState;
        for (int i = 0; i < w.length(); i++)
        {
            String target_edge = Character.toString(w.charAt(i));
            Map<String, State> currentState_edges = this.graph.get(currentState);

            if (currentState_edges.containsKey(target_edge))
                currentState = currentState_edges.get(target_edge);
            else
                return false;
        }

        return this.finalStates.contains(currentState);
    }

    @Override
    public String toString() {

        return String.format("DFA name: %s\n graph:\n %s\n initial state: %s\n final states:\n %s\n",
                this.name, this.graph, this.initialState, this.finalStates);
    }

}

File: python_src/automata/dfa.py

from state import State

class DFA:

    def __init__(self, name:str, alphabet:list,
            graph:dict, initialState:State, finalStates:list):
        
        self._name = name
        self._alphabet = alphabet
        self._graph = graph
        self._finalStates = finalStates
        self._initialState = initialState

    def doesAcceptString(self, w:str) -> bool:

        currentState = self._initialState
        for i in range(0, len(w)):
            target_edge = w[i]
            currentState_edges = self._graph[currentState]

            if target_edge in currentState_edges.keys():
                currentState = currentState_edges[target_edge]
            else:
                return False

        return currentState in self._finalStates

    def __str__(self):

        return "DFA name: %s\n graph:\n %s\n initial state: %s\n final states:\n %s\n" % (self._name, self._graph, self._initialState, self._finalStates)

    def __repr__(self):

        return str(self)
